packages/fields/src/fields/template/template-edit.tsx

```diff
@@ -0,0 +1,245 @@
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useMemo, useState } from '@wordpress/element';
+// @ts-ignore
+import { parse } from '@wordpress/blocks';
+import type { WpTemplate } from '@wordpress/core-data';
+import { store as coreStore } from '@wordpress/core-data';
+import type { DataFormControlProps } from '@wordpress/dataviews';
+
+/**
+ * Internal dependencies
+ */
+// @ts-expect-error block-editor is not typed correctly.
+import { __experimentalBlockPatternsList as BlockPatternsList } from '@wordpress/block-editor';
+import {
+	Button,
+	Dropdown,
+	MenuGroup,
+	MenuItem,
+	Modal,
+} from '@wordpress/components';
+import { useAsyncList } from '@wordpress/compose';
+import { useSelect } from '@wordpress/data';
+import { decodeEntities } from '@wordpress/html-entities';
+import { __ } from '@wordpress/i18n';
+import { getItemTitle } from '../../actions/utils';
+import type { BasePost } from '../../types';
+
+function useTemplates( postType: string ) {
+	return useSelect(
+		( select ) =>
+			select( coreStore ).getEntityRecords< WpTemplate >(
+				'postType',
+				'wp_template',
+				{
+					per_page: -1,
+					post_type: postType,
+				}
+			) ?? [],
+		[ postType ]
+	);
+}
+
+export function useAllowSwitchingTemplates( {
+	postType,
+	postId,
+}: {
+	postType: string;
+	postId: number;
+} ): boolean {
+	return useSelect(
+		( select ) => {
+			const { canUser, getEntityRecord, getEntityRecords } =
+				select( coreStore );
+			const siteSettings = canUser( 'read', {
+				kind: 'root',
+				name: 'site',
+			} )
+				? // @ts-expect-error getEntityRecord is not typed correctly.
+				  getEntityRecord< {
+						page_for_posts: number;
+						page_on_front: number;
+				  } >( 'root', 'site' )
+				: undefined;
+			const templates = getEntityRecords< WpTemplate >(
+				'postType',
+				'wp_template',
+				{
+					per_page: -1,
+				}
+			);
+			const isPostsPage = +postId === siteSettings?.page_for_posts;
+			// If current page is set front page or posts page, we also need
+			// to check if the current theme has a template for it. If not
+			const isFrontPage =
+				postType === 'page' &&
+				+postId === siteSettings?.page_on_front &&
+				templates?.some( ( { slug } ) => slug === 'front-page' );
+			return ! isPostsPage && ! isFrontPage;
+		},
+		[ postId, postType ]
+	);
+}
+
+export const TemplateEdit = ( {
+	data,
+	field,
+	onChange,
+}: DataFormControlProps< BasePost > ) => {
+	const { id } = field;
+	const postType = data.type;
+	const slug = data.slug;
+
+	const allowSwitchingTemplate = useAllowSwitchingTemplates( {
+		postType: data.type,
+		postId: typeof data.id === 'number' ? data.id : parseInt( data.id, 10 ),
+	} );
+	const templates = useTemplates( data.type );
```

Comment:
```suggestion
	const templates = useTemplates( postType );
```
Given you have already defined it up above.